app/src/main/cpp/openxr-program.cc

```diff
@@ -1,19 +1,54 @@
 #include "pch.h"
 
+#include "config.h"
 #include "logger.h"
 #include "openxr-program.h"
 
 namespace zen::display_system::oculus {
 
-bool
-OpenXRProgram::Initialize() const
+namespace {
+
+#define ENUM_CASE_STR(name, val) \
+  case name:                     \
+    return #name;
+
+#define MAKE_TO_STRING_FUNC(enumType)            \
+  inline const char *to_string(enumType e)       \
+  {                                              \
+    switch (e) {                                 \
+      XR_LIST_ENUM_##enumType(ENUM_CASE_STR)     \
+                                                 \
+          default : return "Unknown " #enumType; \
+    }                                            \
+  }
+
+// Create other to_string function on demand
+MAKE_TO_STRING_FUNC(XrResult);
+
+#define TO_STRING(x) #x
+#define MACRO_TO_STRING(x) TO_STRING(x)
+#define FILE_AND_LINE __FILE__ ":" MACRO_TO_STRING(__LINE__)
+
+std::string
+CheckXrResult(XrResult res, const char *originator, const char *source_location)
 {
-  if (!InitializeLoader()) LOG_ERROR("Failed to initialize OpenXR");
-  return true;
+  std::ostringstream err;
+  if (XR_FAILED(res)) {
+    err << "XrResult failure [" << to_string(res) << "]" << std::endl;
+    err << "    Origin: " << originator << std::endl;
+    err << std::string() << "    Source: " << source_location;
+  }
+
+  return err.str();
 }
```

Comment:
This function isn't called from another file, so I suggest move this function to the unnamed namespace.

Reply:
I think it's already in an unnamed namespace.

Reply:
Sorry I overlooked